cpp/src/arrow/flight/flight_test.cc

```diff
@@ -945,6 +945,34 @@ TEST_F(TestFlightClient, DoGetDicts) {
   CheckDoGet(descr, expected_batches, check_endpoints);
 }
 
+TEST_F(TestFlightClient, DoExchange) {
+  auto descr = FlightDescriptor::Path({"ints"});
+  BatchVector batches;
+  auto a1 = ArrayFromJSON(int32(), "[4, 5, 6, null]");
+  auto schema = arrow::schema({field("f1", a1->type())});
+  batches.push_back(RecordBatch::Make(schema, a1->length(), {a1}));
+  std::unique_ptr<FlightStreamReader> reader;
+  std::unique_ptr<FlightStreamWriter> writer;
+  ASSERT_OK(client_->DoExchange(descr, &reader, &writer));
+  ASSERT_OK(writer->Begin(schema));
+  // TODO: Begin() should write the descriptor - this should work even
+  // if we never write any data
+  for (const auto& batch : batches) {
+    ASSERT_OK(writer->WriteRecordBatch(*batch));
+  }
+  ASSERT_OK(writer->DoneWriting());
+  FlightStreamChunk chunk;
+  ASSERT_OK(reader->Next(&chunk));
+  ASSERT_EQ("1", chunk.app_metadata->ToString());
+  ASSERT_OK(reader->Open());
+  AssertSchemaEqual(schema, reader->schema());
```

Comment:
Can you get the schema before reading the first chunk?

Reply:
Yes, that would be equivalent to DoGet essentially - I've added a test case for this.